core/src/test/java/org/apache/brooklyn/enricher/stock/reducer/ReducerTest.java

```diff
@@ -0,0 +1,255 @@
+
+        entity.sensors().set(STR2, "bar");
+        EntityTestUtils.assertAttributeEqualsEventually(entity, STR3, "foo, bar");
+    }
+    
+    @Test
+    public void testReducingBuilderWithJoinerFunctionAndUnusedParameter() {
+        
+        entity.addEnricher(Enrichers.builder()
+            .reducing(StringStringReducer.class, ImmutableList.of(STR1, STR2))
+            .from(entity)
+            .computing("joiner", ImmutableMap.<String, Object>of("non.existent.parameter", "-"))
+            .publishing(STR3)
+            .build()
+        );
+        EntityTestUtils.assertAttributeEquals(entity, STR3, null);
+        
+        entity.sensors().set(STR1, "foo");
+        EntityTestUtils.assertAttributeEqualsContinually(entity, STR3, null);
+
+        entity.sensors().set(STR2, "bar");
+        EntityTestUtils.assertAttributeEqualsEventually(entity, STR3, "foo, bar");
+    }
+    
+    @Test
+    public void testReducingBuilderWithFormatStringFunction() {
+        
+        entity.addEnricher(Enrichers.builder()
+            .reducing(StringStringReducer.class, ImmutableList.of(STR1, STR2))
+            .from(entity)
+            .computing("formatString", ImmutableMap.<String, Object>of("format", "hello, %s and %s"))
+            .publishing(STR3)
+            .build()
+        );
+        EntityTestUtils.assertAttributeEquals(entity, STR3, null);
+        
+        entity.sensors().set(STR1, "foo");
+        EntityTestUtils.assertAttributeEqualsContinually(entity, STR3, null);
+
+        entity.sensors().set(STR2, "bar");
+        EntityTestUtils.assertAttributeEqualsEventually(entity, STR3, "hello, foo and bar");
+    }
+    
+    @Test
+    public void testReducingBuilderWithNamedNonExistentFunction() {
+        try { 
+            entity.addEnricher(Enrichers.builder()
+                .reducing(StringStringReducer.class, ImmutableList.of(STR1, STR2))
+                .from(entity)
+                .computing("unknown function name", ImmutableMap.<String, Object>of("separator", "-"))
+                .publishing(STR3)
+                .build()
+            );
+            Asserts.fail("Expected exception when adding reducing enricher with unknown named function");
+        } catch (Exception e) {
+            Throwable t = Exceptions.getFirstThrowableOfType(e, IllegalStateException.class);
+            Assert.assertNotNull(t);
+        }
+    }
+    
+    private static class Concatenator implements Function<List<String>, String> {
+        @Nullable
+        @Override
+        public String apply(List<String> values) {
+            StringBuilder result = new StringBuilder();
+            for (String value : values) {
+                if (value == null) {
+                    return null;
+                } else {
+                    result.append(value);
+                }
+            }
+            return result.toString();
+        }
+    }
+    
+    public static class StringIntegerReducer extends Reducer<String, Integer> {
```

Comment:
It's worrying that you're defining this in the test - that presumably means it's impossible to do it in yaml, because this class doesn't exist and there is no other way to do it other than writing the code yourself?